### DISCUSSION TITLE
Asynchronous behavioral tests using behave

### DISCUSSION THREAD

**Discussion by petrsynek** [behave/behave#1011] [Q&A] [answered]
**Description of problem:**

I need to write load tests that will fire API calls upon my system, executing multiple asynchronous scenarios. This should be repeated with adjustable "cadency" for a certain period of time simulating the expected system load (data acquisitions, probe messages, client's request ...) and measuring request time, response accuracy, scalability of the system as a whole.

I would like to use BDD semantics so that customer care can write those scenarios while having the option to run multiple scenarios in parallel (asynchronously) either by using some framework or by writing my own scenario launcher (scenario spawner).

**Questions:**
- does Behave support this already?
- is this solvable by repurposing some of the current behave code, e.g: running a single scenario as an async function?

**Detailed vision:**

Example of scenarios:

```
Scenario: Unauthorized user requests data
    Given "random_name" is not in DB # checks that the user is not in DB (user is randomly generated hash)
    Try to access data as "random_name"
    We get a "404" response and log the result # checks the response and logs the outcome

Scenario: Authorized user requests small data
    Given "random_name_2" is in DB # inserts user to DB
    Given "data_small" is in DB # checks if data are in DB or create data
    When we try to access "data_small" as "random_name_2"
    We get a "200" response within "0.1" seconds and log the result # checks the response and logs the outcome

Scenario: Authorized user requests large data
    Given "random_name_3" is in DB
    Given "data_large" is in DB
    When we try to access "data_large" as "random_name_3"
    We get a "200" response within "0.5" seconds and log the result
```

**Example of the functionality:**

For 25 seconds, run scenarios "Authorized user requests large data", "Authorized user requests small data" with a cadency of 20 scenarios per second

**Comment by jenisys** [comment, marked as answer]
## Async Execution

Terminology first: `behave` supports to execute async-steps, but behave normally needs to execute these steps synchronously. The reason for this is that `behave` needs to know the step result before it can continue to the next step (or fail and abort further execution of steps in this scenario). A step writer can use a work-around to dispatch async-step(s) and (wait and) verify the successful execution in the final step(s).

## Parallel Execution of Scenarios

`behave` has currently no official support for a parallel test runner.
There is at least one fork provided by the community (but I am not quite sure how up-to-date the other functionality of `behave` is in this fork compared to the current state of the `behave` repository). See if it works for you or provide your own test runner that does the job.

Otherwise, you could use an external parallel runner of `behave` (basically scripting `behave` subprocess calls as part of a build-script) by using the `file-location` mechanism, like:

```bash
# -- PSEUDO CODE: Execute Scenario in "features/one.feature" at line 42 in parallel by using 10 threads/processes.
$ run-parallel --jobs=10 -- behave features/one.feature:42
```

NOTES:
* Parallel execution of scenarios comes with potential problems (that is independent of `behave`)
* You need to somehow merge the results of the parallel test runs and provide a sane test summary
* This is even more problematic if you repeat to test the same scenario (as in your case of load testing)
* The problem is if a scenario shares a resource, you need to protect it against use in another scenario while in use.
  If you run the same scenario twice (in parallel), you need to solve the problem who is the current owner of the shared resource.
  You may use `behave fixture(s) / fixture-tag(s)` to limit access to a shared resource.
  But that may not be good for load testing.
* **EXAMPLE:** A Scenario modifies a database table. The `database table` are shared resources (same for `database` in some cases).

SEE ALSO:

* https://github.com/hugeinc/behave-parallel (or look at `pypi.org`)

## Load Testing

Much of what you need (and have mentioned above) is normally provided by your load test runner.
It used `multi-mechanise` in the past. There may be better load testing tools by now.
It had the functionality to repeat a certain test / scenario in parallel for predefined durations.
You could also mix different loads and had a nice post-processing of the results as graphical diagram.

SEE ALSO:

* https://github.com/cgoldberg/multi-mechanize
* https://github.com/jenisys/multi-mechanize

**Reply by petrsynek** [reply]
Thanks for so exhaustive answer!

**Maybe one last question:** 
Each scenario is basically a set wrapper of functions parsed from the scenario description, is there any simple way how to extract this functionality from behave? (which is what I need) 

To be explicit is there any simple way to extract this scenario -> set of functions functionality out of behave? 
E.g: call something like (this is just an untested example if you spot an error just ignore it)

```
import asyncio
import behave

scenarios: Dict[str :List[callable]] = behave.gather_scenarios() # e.g: returns dics of scenario name : list of individual steps

async def scenario_run(steps: List[callable]):
    for step in steps:
        await step()

async def main():
    for scenario in scenarios:
        asyncio.create_task(scenario_run(scenario_run())) # rund scenario in background

if __name__ == '__main__':
  asyncio.run(main())
```

**Comment by jenisys** [comment]
TERMINOLOGY FIRST:

* Step: Represents the a step in a Feature file, basically its text line, parameters, text and step table parts
* Step Implementation (function): Represents the function that is executed when a step is "executed"; often an "anonymous function" (like: Many may have the same function-name, like "step_impl()")

`behave` provides a data model for its entities / idioms.
The core functionality of the data model is implemented in the `behave.model` module.
The Gherkin parser (in module `behave.parser`) parses Feature files (or parts of these files, like: Scenario part or steps of a Scenario) and creates the corresponding data model objects from it.
These object trees are executed a test runner that may selected a subset of Features and/or Scenarios to be executed.

DATA MODEL CORRESPONDENCE:

* Feature: `behave.model.Feature` class
* Scenario: `behave.model.Scenario` class
* Step: `behave.model.Step` class
* Step Implementation function: `behave.model.Match` class as proxy/adapter to run/call the Step Implementation function (with its parsed and optionally type-converted parameter values)

Otherwise, the `step_registry` (module: `behave.step_registry.StepRegistry`) contains the registered / known step-functions (Step Implementation functions) after the steps are loaded by the test runner.

`behave` has an synchronous execution model (and not an asynchronous one).
Because execution of steps has a run-to-completion semantic, an async model has also no real benefit (currently).
The `@step` decorators (and `@given`, `@when`, `@then`) provide glue code if the step implementation function is an asynchronous coroutine. Therefore, the coroutine is still executed in synchronous way.

NOTES:
* Your pseudo-code above is overly simplistic (hook processing, error handling, etc. are missing)

APPROACHES TO RUN (ASYNC) STEP EXECUTION:
* Provide your own test-runner (as plugin) for `behave`. This functionality is already contained in the current `main` branch of the Git repository. Some minor details may change in the near feature, but it is already usable.
* Use the Gherkin parser to parse Feature files or Scenarios (as text) parts. This provides you with data model elements, like iterating over the.steps of a scenario and "running them" (like via `Scenario.run()` or `Step.run()`)
* Use `behave.runner.Context.execute_steps(steps_as_text)` to execute some steps as text (gets parsed and executed)
